ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/genesis/GenesisHandler.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.statetransition.genesis;
+
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.isThereEnoughNumberOfValidators;
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.is_valid_genesis_state;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.operations.DepositWithIndex;
+import tech.pegasys.artemis.datastructures.state.BeaconStateWithCache;
+import tech.pegasys.artemis.datastructures.util.DepositUtil;
+import tech.pegasys.artemis.datastructures.util.GenesisGenerator;
+import tech.pegasys.artemis.pow.api.DepositEventChannel;
+import tech.pegasys.artemis.pow.api.MinGenesisTimeBlockEventChannel;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.pow.event.MinGenesisTimeBlockEvent;
+import tech.pegasys.artemis.statetransition.events.GenesisEvent;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class GenesisHandler implements DepositEventChannel, MinGenesisTimeBlockEventChannel {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ChainStorageClient chainStorageClient;
+  private final GenesisGenerator genesisGenerator = new GenesisGenerator();
+
+  private Queue<DepositsFromBlockEvent> bufferedDepositsFromBlockEvents = new LinkedList<>();
+  private volatile boolean genesisAlreadyTriggered = false;
+
+  public GenesisHandler(final ChainStorageClient chainStorageClient) {
+    this.chainStorageClient = chainStorageClient;
+  }
+
+  @Override
+  public synchronized void onDepositsFromBlock(final DepositsFromBlockEvent event) {
+    if (!chainStorageClient.isPreGenesis() || genesisAlreadyTriggered) {
+      return;
+    }
+
+    if (isThereEnoughNumberOfValidators(genesisGenerator.getCandidateState())) {
+      bufferedDepositsFromBlockEvents.add(event);
```

Comment:
Shouldn't we process this immediately if we already know the min genesis block and know this block is before it?